diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -592,6 +592,11 @@ set_SIMULATeQCD_property(hisqForce PROPERTIES RUNTIME_OUTPUT_DIRECTORY "testing")
 SIMULATeQCD_target_compile_definitions(hisqForce PRIVATE HALODEPTHSPIN_4=1 HALODEPTH_0=1 SINGLEPREC=1 DOUBLEPREC=1 COMP_R18=1 COMP_R14=1 COMP_U3R14=1 NSTACKS_1=1 NSTACKS_12=1 NSTACKS_14=1 ${ALL_LAYOUTS})
 add_to_compound_SIMULATeQCD_target(tests hisqForce)
 
+set_SIMULATeQCD_gpu_backend(src/testing/main_hisqForceBenchmark.cpp ${SOURCE_FILES_DSLASH} src/modules/inverter/inverter.cpp src/modules/hisq/hisqForce.cpp)
+add_SIMULATeQCD_executable(hisqForceBenchmark src/testing/main_hisqForceBenchmark.cpp ${SOURCE_FILES_DSLASH} src/modules/inverter/inverter.cpp src/modules/hisq/hisqForce.cpp)
+set_SIMULATeQCD_property(hisqForceBenchmark PROPERTIES RUNTIME_OUTPUT_DIRECTORY "testing")
+SIMULATeQCD_target_compile_definitions(hisqForceBenchmark PRIVATE HALODEPTHSPIN_4=1 HALODEPTH_0=1 SINGLEPREC=1 DOUBLEPREC=1 COMP_R18=1 COMP_R14=1 COMP_U3R14=1 NSTACKS_1=1 NSTACKS_12=1 NSTACKS_14=1 ${ALL_LAYOUTS})
+
 set_SIMULATeQCD_gpu_backend(src/testing/main_hisqSmearingTestMulti.cpp src/spinor/spinorfield.cpp)
 add_SIMULATeQCD_executable(hisqSmearingMulti src/testing/main_hisqSmearingTestMulti.cpp src/spinor/spinorfield.cpp)
 set_SIMULATeQCD_property(hisqSmearingMulti PROPERTIES RUNTIME_OUTPUT_DIRECTORY "testing")
diff --git a/src/base/wrapper/gpu_wrapper.h b/src/base/wrapper/gpu_wrapper.h
--- a/src/base/wrapper/gpu_wrapper.h
+++ b/src/base/wrapper/gpu_wrapper.h
@@ -11,7 +11,7 @@
 #include <cuda_runtime_api.h>
 #include <cuda_fp16.h>
 #include <utility>
-#include "nvToolsExt.h"
+#include <nvtx3/nvToolsExt.h>
 
 
 #define gpuError_t                       cudaError_t
